rfc/rfc-43/rfc-43.md

```diff
@@ -0,0 +1,222 @@
+
+### Request Handler
+
+Receive the client's request and save the plan in the meta table.
+
+### Meta Table
+
+Meta table has been implemented by using an internal HUDI MOR Table to store the required metadata for the compaction/clustering message. This table will be internal to a dataset and will not be exposed directly to the user to write/modify.
+
+The Meta Table has the following features:
+
+1. Is a HUDI MOR Table
+
+2. Defines its own custom payload HoodieServicePayload
+
+3. Use basePath + instant as recordKey to guarantee unique
+
+4. Follows all the conventions of a HUDI Table
+
+### Scheduler
+
+Periodically scan the Meta Table and submit compaction/clustering job according to user-specified rules, and need to plug-in the execution engine.
+
+### Schema
+
+Being an actual Hoodie Table, the Meta Table needs to have a schema to store the compaction/clustering message. 
+
+```
+{
+    "namespace":"org.apache.hudi.avro.model",
+    "type":"record",
+    "name":"HoodieServiceRecord",
+    "doc":"Compaction/Clustering message saved within the Meta Table",
+    "fields":[
+        {
+            "name":"table_name",
+            "type":"string"
+        },
+        {
+            "name":"base_path",
+            "type":"string"
+        },
+        {
+            "name":"instant",
+            "doc":"compaction/clustering instant for the hoodie table",
+            "type":"string"
+        }, {
+            "name":"action",
+            "doc":"compact or cluster for this instant",
+            "type":"string"
+        },{
+            "name":"status",
+            "doc":"Status of the compaction/clustering job, 0：Scheduled 1：Running 2：Retrying 3:  Completed",
+            "type":"int"
+        } ,{
+            "name":"isDeleted",
+            "doc":"True if this plan has been deleted",
+            "type":"int"
+        }
+    ]
+}
+```
+
+## Interface design
+
+### Register
+
+/v1/hoodie/service/register
+
+```
+{
+"table_name":"table_name",
+"base_path":"/hoodie/base_path",
+"compaction_config":"compaction_config",
+"clustering_config":"clustering_config"
+}
+```
+
+### Schedule
+
+/v1/hoodie/service/schedule
+
+```
+{
+"table_name":"table_name",
+"base_path":"/hoodie/base_path",
+"action":"compact/cluster",
+"instant":"202111161559"
+}
+```
+
+
+### Delete
+
+/v1/hoodie/service/delete
+
+```
+{
+"table_name":"table_name",
+"base_path":"/hoodie/base_path",
+"action":"compact/cluster",
+"instant":"202111161559"
+}
+```
+
+## Error Handling
+
+Due to two tables being involved in each operation, we need to ensure that they are in sync and errors during dataset operation are handled correctly.
+
+1. Client scheduled the plan but request failed.
+
+2. RequestHandler received request but the commit is not completed.
+
+3. Client rollback plan after request to Compaction/Clustering Service.
```

Comment:
by client we mean regular writer to original hudi table right. Let's try to keep the overhead less as possible. 

I guess here we are employing push based mechanism. Alternatively, can we think a pull based one. Each hudi table will register w/ this service for table services it likes to delegate. for eg, tableA can delegate just cleaning, tableB can delegate cleaning and compaction. 
Once registered, regular writers don't make any further communication w/ the service. The request handler will periodically poll the datatable timeline of these tables and check for any requested instants for table services and work on them. i.e. update the internal meta table, etc. 
This way, the client does not need to worry about whether request handler is available or not. if not available, should it abort and continue it inline etc.

Reply:
This is a good proposal. Such a design will be more flexible. I previously considered that regular polling of the service may concentrate the pressure on the service and cause delays. Performance optimization can be placed in follow up.

Reply:
The pull based mechanism works for less number of tables. Scanning 1000s of tables for possible services is going to induce lots of load of listing.

The push based model is actually very elegant. Let me explain how I envision it:

Assume a regular table which has some cleaning setting. In the current code, we schedule the clean and then execute the clean in the same Spark application. When hoodie.service.enable=true, the scheduling of clean will be as earlier but the execution part (HoodieWriteClient.clean(...) will detect that TableServices are enabled and instead of running clean, it will schedule it with the TableService by calling in the API endpoint. 

Benefits:
1. Easy Integration with HUDI current functionality
2. No extra load to list datasets to find operations to run
3. Confirms to all HUDI write config settings (e.g. when to clean, when to compact)
4. In case of downtime of TableServices, one can simply disable hoodie.service.enable and revert to inline functionality.

Reply:
Furthermore, we can implement a Java client for table services - HoodieTableServiceClient - which can enable scheduling any type of services from within the logic of table operation.

E.g. Assume the HoodieWriteConfig enable record level index. Since the index does not exist yet, it needs to be bootstrapped. There are two possibilities here:
1. The user manually runs async-indexing to initialize the record index
2. The metadata table code detects that record index is enabled but not present and calls HoodieTableServiceClient.scheduleService(...) to enable async indexing automatically.

The 2 scenario above can be extended to various intelligent cases:
1. The ingestion side code decided when it is time to compact the MOR table (based on some metric like read time / write time) rather than a fixed schedule like after 10 deltacommits.

Reply:
+1 on push-based model. The fallback to inline is covered in push model.